pkg/scheduler/queue/queue.go

```diff
@@ -0,0 +1,122 @@
+type PolicySnapshotKeySchedulingQueueWriter interface {
+	Add(cpsKey PolicySnapshotKey)
+}
+
+// SchedulingQueue is an interface which queues PolicySnapshots for the scheduler to schedule.
+type PolicySnapshotKeySchedulingQueue interface {
+	PolicySnapshotKeySchedulingQueueWriter
+
+	// Run starts the scheduling queue.
+	Run()
+	// Close closes the scheduling queue.
+	Close()
+	// NextClusterPolicySnapshotKey returns the next-in-line PolicySnapshot key for the scheduler to schedule.
+	NextClusterPolicySnapshotKey() PolicySnapshotKey
+	// Done marks a PolicySnapshot key as done.
+	Done(cpsKey PolicySnapshotKey)
+}
+
+// simplePolicySnapshotKeySchedulingQueue is a simple implementation of PolicySnapshotKeySchedulingQueue.
+//
+// At this moment, one single workqueue would suffice, as sources such as the cluster watcher,
+// the binding watcher, etc., can catch all changes that need the scheduler's attention.
+// In the future, when more features, e.g., inter-placement affinity/anti-affinity, are added,
+// more queues, such as a backoff queue, might become necessary.
+type simplePolicySnapshotKeySchedulingQueue struct {
+	policySanpshotWorkQueue workqueue.RateLimitingInterface
+}
+
+// Verify that simplePolicySnapshotKeySchedulingQueue implements PolicySnapshotKeySchedulingQueue
+// at compile time.
+var _ PolicySnapshotKeySchedulingQueue = &simplePolicySnapshotKeySchedulingQueue{}
+
+// simplePolicySnapshotKeySchedulingQueueOptions are the options for the simplePolicySnapshotKeySchedulingQueue.
+type simplePolicySnapshotKeySchedulingQueueOptions struct {
+	workqueueRateLimiter workqueue.RateLimiter
+	workqueueName        string
+}
+
+// Option is the function that configures the simplePolicySnapshotKeySchedulingQueue.
+type Option func(*simplePolicySnapshotKeySchedulingQueueOptions)
+
+var defaultSimplePolicySnapshotKeySchedulingQueueOptions = simplePolicySnapshotKeySchedulingQueueOptions{
+	workqueueRateLimiter: workqueue.DefaultControllerRateLimiter(),
+	workqueueName:        "policySnapshotKeySchedulingQueue",
+}
+
+// WithWorkqueueRateLimiter sets a rate limiter for the workqueue.
+func WithWorkqueueRateLimiter(rateLimiter workqueue.RateLimiter) Option {
+	return func(o *simplePolicySnapshotKeySchedulingQueueOptions) {
+		o.workqueueRateLimiter = rateLimiter
+	}
+}
+
+// WithWorkqueueName sets a name for the workqueue.
+func WithWorkqueueName(name string) Option {
+	return func(o *simplePolicySnapshotKeySchedulingQueueOptions) {
+		o.workqueueName = name
+	}
+}
+
+// Run starts the scheduling queue.
+//
+// At this moment, Run is an no-op as there is only one queue present; in the future,
+// when more queues are added, Run would start goroutines that move items between queues as
+// appropriate.
+func (sq *simplePolicySnapshotKeySchedulingQueue) Run() {}
+
+// Close shuts down the scheduling queue.
+func (sq *simplePolicySnapshotKeySchedulingQueue) Close() {
+	sq.policySanpshotWorkQueue.ShutDownWithDrain()
```

Comment:
IIRC, this call will be blocked by existing reconcile activities. Shall we have a non-blocking call too?

Reply:
Hi Ryan! Yeah, this call will block until all work items are Done()'d. I will add a non-blocking call as well.